### ISSUE TITLE
Task 2.0.0 Feedback

### ISSUE THREAD

**Issue by dbaeumer** [Microsoft/vscode#28235]
We did a Task 2.0 deep dive today and this issue collects the combined feedback received. 

### Task Properties

- echo should be true by default so that users get a better idea of the task executed
- we should support flattening the JSON properties as we do with other settings (e.g `"terminal.echo": true` instead of `"terminal": { "echo": true }`

**Comment by dbaeumer** [comment]
@Microsoft/vscode please read. Feedback and corrections welcome.

**Comment by Tyriar** [comment]
> We also discussed the need of a task viewlet but decided that this will not add any big additional value over the task picker and a separate panel.

@Lixire and @aefernandes are currently looking into this I believe.

**Comment by Tyriar** [comment]
Most of this is In response to:

> It was suggested to run tasks in its own panel instead of reusing the terminal panel.

First off, it definitely looks like these sorts of improvements will make tasks more useful and usable. I just wonder if this is the right direction we should be taking:

- We already have output, debug console and terminal which are all very similar.
- We don't deal with overflowed panel actions too well right now, this will make the panel less usable on small monitors and/or high scale/zoom settings. This is what the panel looks like right now no my 13" laptop screen. A new panel may start to squash the panel actions off screen with this window position:
  ![image](https://user-images.githubusercontent.com/2193314/26941756-7774159a-4c34-11e7-94ed-79a20096cfb8.png)
- Are tasks generally useful enough to warrant their own panel? Personally I've found little benefit over shell scripts, particularly if you want the project to be portable as tasks are only usable in VS Code.
- Perhaps time invested in tasks would be better spent bringing some of these sorts of features to the terminal, namely https://github.com/Microsoft/vscode/issues/20676 + https://github.com/sourcelair/xterm.js/issues/576 which could lead to exit code marking, folding of terminal commands and clever problem matching (leveraging settings system for definitions).

### Technical details

On the implementation side, this is definitely technically possible. It will put a permanent halt on moving tasks to an extension, I think that effort has fizzled out anyway though.

`TerminalInstance` is fairly generic right now and it wouldn't require that much tweaking to get working inside another panel. It doesn't have a dependency on `TerminalService` and communicates with it via events. The main problems you will run into are how terminal keybindings and the terminal context keys interact with the new tasks system:

- `terminalTextSelected` is fairly easy, this could be changed to be if text is selected in tasks/terminal and that panel is active. 
- `terminalFocus` is a little trickier as making this apply to both task panel focus and terminal panel focus would cause issues with my keybindings, for example:

  ```js
  { "key": "ctrl+`", "command": "workbench.action.terminal.focus",
                     "when": "!terminalFocus" }
  ```

I don't think too much of `TerminalService` can be reused if we make a tasks panel.

### Some alternative ideas:

- We implement terminal tabs and each task gets its own tab https://github.com/Microsoft/vscode/issues/10546
- We embed `TerminalInstances` (ie. xterm.js) within the output panel, disabling stdin by default, then move tasks back to output?
- Refocus effort on tasks on improving the terminal

**Comment by dbaeumer** [comment]
See #28351

**Comment by dbaeumer** [comment]
Added support for MRU list:

![capture](https://user-images.githubusercontent.com/1931590/26992991-7aa4a814-4d61-11e7-9eaa-8e4749972f45.png)

**Comment by alexandrudima** [comment]
@Tyriar I don't think the idea of having a Task panel competes with your mentioned improvements for Terminal. If there is a UX path where Tasks and Terminal can fit in the same panel, that would also be great.

Some reasons why IMHO a Task panel would help:
* It would be nice to see somewhere, in the form of a list, the current running tasks and some sort of history with past ran tasks. From the list, it would be nice to be able to expand output for a task in case there is a problem (sort of like travis).
* Such a list would be especially helpful for managing **background tasks**
* A panel might help with **task discoverability**. At this time, there is no obvious UI path to go launch a task or configure a task from a terminal. There is no mention of tasks or the possibility of gathering errors from a task in the UI. 

And what would also be very nice:
* It would be nice that typing in the terminal auto-completes with information from the task providers. e.g. typing `"gulp c"` should complete with the gulp tasks that start with letter `c`.
* It would be nice that running a command in a terminal can be configured into a task via a one-click. e.g. it would be nice if, when someone runs gcc in the terminal (without a task configured) that we can detect that this would be a prime candidate for a task and we would guide the user into creating a task. I think many folks don't know that a command they run in the terminal can be "lifted" to be even more integrated into VS Code.

**Comment by bpasero** [comment]
I would still want us to explore the possibility of a panel for tasks, I think UX scalability as reason to not do it is not very fair, we need to solve the panel overflow no matter what in the same way we solved it for the activity bar (being able to show/hide panels, reorder them, see a "..." overflow menu when space is too little).

I do agree though there is some overlap in Debug Console, Integrated Terminal, Output and Tasks, however I am not sure if it would make sense to put them all into one. There is very specific scenarios for each of the panels:
* output: after tasks moved to terminal I would argue output is only ever being used as a way to see diagnostics (I wonder if we should make output less prominent)
* debug console: has rich functionality when debugging, but otherwise is quite useless (maybe we should only show it when you are in a debug session?)
* terminal: this is where I execute shell commands. nothing in here is related to VS Code, it is just convenient to not have to leave VS Code to use it
* tasks: this is VS Code specific: I execute potentially multiple long running tasks, each with output and progress and possible error conditions (markers)

Here is a mockup for the Tasks panel:

![image](https://user-images.githubusercontent.com/900690/27125029-44ab15f0-50f3-11e7-8fd7-a6e8a5263029.png)

We can provide an explicit configure action per task which would guide users to configure problem matchers.

**Comment by dbaeumer** [comment]
@bpasero thanks for the nice mockup.

**Comment by jrieken** [comment]
New ideas wrt task customisation and referencing (contributed) tasks by an id or something else? I think from an API POV that is a thing to tackle.

**Comment by chrmarti** [comment]
Not being an active task user myself, I wonder if the suggested Tasks panel could also show processes run from the Terminal. That could then open a path to offer problem matchers for tools run in the Terminal.

**Comment by roblourens** [comment]
The mockup omits the problems panel. I wonder whether that could be rolled into the tasks panel to save space?

> _debug console_: has rich functionality when debugging, but otherwise is quite useless (maybe we should only show it when you are in a debug session?)

Hiding it when a debug session hasn't been run would be useful. After a debug session completes, I still want to examine the contents of the console though.

**Comment by rebornix** [comment]
As I'm not an active task user but I'm also thinking about why I don't use Task too often

* If no one sets up all the tasks in a project, I don't think I'll ever know this feature from daily use. To new users, you don't have an idea when the Task will get in the way. A **task viewlet or panel** can help with discoverability in this case.
* I didn't know where to launch/kill the task until one day I happen to type task in the command palette. Task 2.0 uses Terminal which helps with the problem a little as I can kill the task the same way as a shell cmd.
* Task 2 now uses Terminal for emitting output, the first time experience was not intuitive as it took over the Terminal even though I was using it. Besides it shares the same persistence problem as Terminal, after reloading the window I no longer have access to background tasks.

Speaking of debug and output panel

* I'm not a fan of auto show/hide unless we make it correct.
  a. If we have a clean state of the workspace, launch debugger, the debug panel would show if there is no prelaunch task and no output.
  b. If we have a prelaunch task, the terminal shows up first as it provides the task info the users but after the task is executed successfully, the debug panel doesn't show up.
  c. I don't have a good reproduce of step a and b. I don't have a good expectation which will show up last.
  d. Last time I get the same confusion is the competence of Output panel and Debug panel

Lastly, do you think it make sense to emit output directly terminal instance the same way as Task 2? The only thing you can do in the output panel is view/copypaste/search and all of them are doable in Terminal, the discoverability won't be worse than current Task 2.

**Comment by dbaeumer** [comment]
@jrieken the initial description contains what we discussed in terms of how to reference a task (see Task Customization). I haven't started implementing anything. First need to see how feasible that is. Any comments to the proposal in terms of API feedback.

**Comment by irekeri** [comment]
As a heavy user of the task/debug infrastructure within Visual Studio Code, I felt that it might be worth it to document my observations with the existing system.

My development environment is heavily focused on testing which has caused a large number of test executables to be created along side the actual application being developed. Additionally, all code compilation as well as test executable execution happens remotely (Virtual Studio Code running on OS X, where as compilation / test executable happen on a Linux machine).

I have found tasks to be very flexible and with the most recent additions to the Insiders Build, building and executing tests with a few keystrokes has really streamlined my development environment. I have been able to augment the tasks I execute today with a simple SSH script used to bridge the gap between Virtual Studio Code and the remote machine where all of the compilation happens.

That being said, one thing I have noticed is that I find myself duplicating task and launch configurations to handle the number of executables I deal with on a day-to-day basis. Having both my tasks.json and launch.json filled with upwards of 30+ configurations only differing by the executable is a little cumbersome. Additionally, the test executables themselves take arguments relating to specific filters of which subtests to actually execute (which is very helpful in executables which contain hundreds of tests). This means having to open up and edit my tasks.json any time I want to execute a smaller subset of a test executable.

While it is not clear to me what belongs in an extension versus what belongs in the core functionality of the editor, some things I would like to see are:
- The ability to use 1 task configuration as a template, and feed in an array variations on that same template to help curb the duplication of task configurations.
- The ability to specify a command (i.e. ${command:getDynamicArgument}) to be used to prompt a user for additional arguments on a task (which would help make it more fluid in scenarios where the arguments to an executable are constantly changing). Additionally a history of recent inputs would really improve the user experience of it.
- Something similar to pipeTransport within the cpptools C++ extension which would allow the remote execution of a task.

I understand that some of these items are (and I have solved them via script development) achievable via external scripts, but I see these as problems that numerous developers across all walks of life would run in to and most likely solve in very similar ways.

**Comment by dbaeumer** [comment]
We again discussed the tasks.json schema today and came to the following conclusion.

- there shouldn't be a separate customization section. So everything goes below the tasks section
- there should be a clear visual distinction between defined tasks and task customizations.

Below is a proposal how this will look in tasks.json
```ts
{
    "version": "2.0.0",
    "tasks": [
        // This is a defined tasks not contributed by an extension
        {
            "type": "raw",
            "name": "yeoman",
            "command": "yo",
            "runtime": "shell",
            "group": "build",
            "presentation": {
                "reveal": "always",
                "focus": false,
                "panel": "dedicated"
            }
        },
        // This is a customization of the npm build script
        {
            "type": "npm",
            "script": "build",
            "configure": {
                "problemMatcher": ["$tsc"]
            }
        },
        // This is a npm script task which the extension needs to
        // fully resolve. It will execute the osx and the osx-min
        // script
        {
            "type": "gulp",
            "tasks": ["osx", "osx-min"]
        }
    ]
}
```

**Comment by Ciantic** [comment]
This has a weird title. But here is my Issue with v.2.0.0:

If you run a background task, and close the VSCode, the background task is left running. This is quiet annoying.

Is this feedback issue meant for this kind of feedback? Thanks.

**Comment by Ciantic** [comment]
This is unrelated to my bug above, but I would like to see tasks having own tabs here:

![image](https://user-images.githubusercontent.com/64731/27988195-d4a2f3e6-6424-11e7-8e69-1dc4a29c23cf.png)

If I could see e.g. "build" or "testwatch" as a tab there, it would be more accessible and easier to notice it's running those tasks. The dropdown is too subtle for me, but this is matter of taste.

**Comment by Tyriar** [comment]
@Ciantic terminal processes leaking tracked in https://github.com/Microsoft/vscode/issues/26807. See https://github.com/Microsoft/vscode/issues/28235#issuecomment-308374308 for proposal on a tasks tab.

**Comment by aefernandes** [comment]
Thoughts on what we (+ @Lixire) seem to expect from a task panel and current mockups. Feedback on brain dump and mockups welcome.

- The panel should show a summary of each completed task, including execution time, exit code, and the foldable output if applicable. By default, we would show tasks in MRU order and prioritize tasks that are currently running. This would also include available/auto-detected tasks. 
- Allow user to start, stop, or (if already run) restart a task with one click. 
- I would expect the tasks panel to open from the newly-added Tasks menu and from the command palette. However, other panel items open from the View menu - we could include a Tasks option there as well.
- We should provide guidance if tasks have not been configured and panel is open. Perhaps a link to the documentation, or a single button that configures .
- Provide an explicit configure action for each task, and an action to configure problem matchers when viewing output. 
- A stretch goal could be to allow users to configure or make quick edits without ever touching JSON (about 40% of users interviewed indicated JSON configuration was a barrier to usage).
- To support multi root workspaces, an option is to show a filepath next to each task name similar to the way debug works. 

Example UI options:
A) ![task panel b](https://user-images.githubusercontent.com/7109548/28031444-31587b00-655c-11e7-8e2f-47bc6378b5b0.png)
B) ![task panel a](https://user-images.githubusercontent.com/7109548/28031442-2f29a3b8-655c-11e7-9039-7115c9c820f4.png)
C) ![task panel c](https://user-images.githubusercontent.com/7109548/28031448-34e5e5f0-655c-11e7-919b-25af0387bf4f.png)

**Comment by aefernandes** [comment]
Thoughts on what we (+ @Lixire) seem to expect from a task panel and current mockups. Feedback on brain dump and mockups welcome.

- The panel should show a summary of each completed task, including execution time, exit code, and the foldable output if applicable. By default, we would show tasks in MRU order and prioritize tasks that are currently running. This would also include available/auto-detected tasks.
- Allow user to start, stop, or (if already run) restart a task with one click.
- I would expect the tasks panel to open from the newly-added Tasks menu and from the command palette. However, other panel items open from the View menu - we could include a Tasks option there as well.
- We should provide guidance if tasks have not been configured and panel is open. Perhaps a link to the documentation, or a single button that configures .
- Provide an explicit configure action for each task, and an action to configure problem matchers when viewing output.
- A stretch goal could be to allow users to configure or make quick edits without ever touching JSON (about 40% of users interviewed indicated JSON configuration was a barrier to usage).
- To support multi root workspaces, an option is to show a filepath next to each task name similar to the way debug works.

**Example UI options:**

A) Task list with output opening in sidebar
![task panel a](https://user-images.githubusercontent.com/7109548/28036916-a1d27c66-656e-11e7-8eb3-bc239f5daf85.png)

B) Task list in MRU order
![task panel c](https://user-images.githubusercontent.com/7109548/28036927-a8d75810-656e-11e7-8115-e974938ba599.png)

C) Task list with groupings
![task panel b](https://user-images.githubusercontent.com/7109548/28036939-b1069dca-656e-11e7-8c61-c17b990e8679.png)

**Comment by Ciantic** [comment]
I'd like to point out that in my opinion it's a mistake to advertise this as default option for VSCode stable (1.14.0), if even the termination of the tasks does not work. https://github.com/Microsoft/vscode/issues/26807

**Comment by dbaeumer** [comment]
Just to clarify things: the termination without leakage doesn't work for npm tasks. It does work for all others kind of tasks.

**Comment by gep13** [comment]
@dbaeumer I have a question/suggestion about this new Task Window...

When in Visual Studio, using the Task Runner Explorer, it is possible to click on one or more of these configurable buttons:

![image](https://user-images.githubusercontent.com/1271146/28367639-9ef9862a-6c89-11e7-939e-994a9bbd338a.png)

These then append additional arguments onto the task that is being executed.  Is it possible to do something similar here?  In the screenshot above, I am selecting to add the `Verbose` flag onto the task so that additional information is added to the output log.

**Comment by Jeyanthinath** [comment]
It would be great if vscode would support **pycharm style** , task configure and task runner on the top right with run button, 

Suppose user need to run multiple configuration file for same file , he can very well setup and need to go to  command pannel for task running ,ther is **no **one click solution****

**Comment by nonameolsson** [comment]
Love this idea! I also like the screenshot https://github.com/Microsoft/vscode/issues/28235#issuecomment-316376605. This reminds me also how JetBrains manage their panels, with clickable buttons etc. 
You can see an idea of what I refer to here https://github.com/Microsoft/vscode-react-native/issues/481

So far VS Code shows text in the panels. Would it be possible to extend it with buttons, lists etc? That would make the panels more interactive.

**Comment by rkeithhill** [comment]
@gep13 - that's more or less what I'd like to see.  Having `Tasks` show up in the `Side Bar` is where they belong in my opinion.  It elevates the Tasks concept to the proper level.  What should be displayed in the Output/Debug/Terminal area is task output from various task executions.

**Comment by nonameolsson** [comment]
So I have been thinking about how I would like it to be. As I already [mentioned](https://github.com/Microsoft/vscode/issues/28235#issuecomment-317287321) I like the solution from JetBrains and from Visual Studio. 

Here is a suggestion. What do you think? Of course there are things I haven't figured out yet, but it's a first draft suggestion at least.

![vscode tasks 2 0](https://user-images.githubusercontent.com/1493018/28836342-1f7b2032-76ae-11e7-8e48-ea28e7ebd96e.jpg)

**Comment by rkeithhill** [comment]
@nonameolsson I very much like the Tasks panel in the Side Bar.  I'm not sure you really need the "RUNNING TASKS" view given that the Tasks panel in the Side Bar can indicate (via icon or text) which tasks are currently running.  And selecting a task from there could switch the tasks output view to the output of the selected task.  Everything else looks really good to me!

**Comment by nonameolsson** [comment]
@rkeithhill That is a good suggestion! I like it, that way the UI will be a lot cleaner. Maybe a checkbox in the task panel to show only running tasks?

**Comment by nonameolsson** [comment]
Update based on suggestion in https://github.com/Microsoft/vscode/issues/28235#issuecomment-319429369

![vscode tasks 2 0 version 2](https://user-images.githubusercontent.com/1493018/28854453-c942b62c-76fc-11e7-92d2-7abc015422d0.jpg)

**Comment by dragonwolf83** [comment]
I agree, moving Tasks to Activity Bar is the perfect place for it to be. 

With that change, I don't think you need a separate Task Panel. The Tasks can remain in Terminal Panel and work on [Tabs for Terminal #10546](#10546) and [Split Terminal #7504](#7504) to gain the same benefits while extending it to popular use cases.

**Comment by gep13** [comment]
In the screenshot that I showed (https://github.com/Microsoft/vscode/issues/28235#issuecomment-316376605), the icons down the left hand side are not tasks, they are additional, optional, parameters that can be passed to the selected task.  I would still like to see this ability included as well.

**Comment by Jeyanthinath** [comment]
Love @nonameolsson suggested UX 

non ux suggession: 

**Adding peak memory usage for each task can be also added** gives much controll over each task making the vscode to run smooth

**Comment by nonameolsson** [comment]
@Jeyanthinath Thanks!

@dragonwolf83 Great suggestions regarding tabs for terminal and split terminal also! That would be awesome! I would love to be able to see multiple running tasks at the same time in a split terminal.

@gep13 Are those icons general, or for each task? In other words: do you choose verbose for a task before starting it, or does it affect all tasks?

**Comment by gep13** [comment]
>@nonameolsson said...
Are those icons general, or for each task? In other words: do you choose verbose for a task before starting it, or does it affect all tasks?

Not in a position to test just now, but I believe that once selected those buttons affect each task that is ran.  i.e. you enable/disable them as required as you are executing tasks.

**Comment by nonameolsson** [comment]
Any update on this? I really think that improvements to the Task section could make VS Code better.

**Comment by dbaeumer** [comment]
No, our current focus it to support multi root for Tasks.

**Comment by jeyanthimuthuram** [comment]
Any Updates on this ? I hope now multi-roots are into stable build .

**Comment by nonameolsson** [comment]
What is the status of this?

**Comment by dbaeumer** [comment]
@nonameolsson the technical features for task 2.0.0 got implemented. However no work has started yet to build a special task UI. However we added proposed API for extensions to query and execute tasks, so that task providers can add special UI (for example npm can contribute actions to the package.json file to directly execute scripts from the explorer).

**Comment by Guema** [comment]
@bpasero @nonameolsson borth your ideas seems really cool & mandatory to me.

I personnally need this feature, and Imo this is critical.
What i lack is mostly an overview panel of running / available tasks.
This list ui should stay visible when looking at a particular task log.
It should be like tabs/folds you can click to see each one details.
Terminal dropdown is not so good at it.

**Comment by Ciantic** [comment]
I like the @nonameolsson GUI, however I would add it that running tasks should be shown as tabs in the panel so that toggling between the outputs is faster.

![tasks-with-tabs](https://user-images.githubusercontent.com/64731/40735757-ab1ee8f6-6444-11e8-90df-c0828750c665.png)

My assumption is that there is very few running tasks and tabs are easy way to distinguish between the running task outputs.

**Comment by gulshan** [comment]
I had a simple proposal for task discovery- show configured tasks under "Task" menu. #47172

**Comment by nonameolsson** [comment]
I really would like to see some improvements to the task session. However, I'm really used to use the terminal for all that now. And being able to split into two terminals at the same time in VS Code has been a big help for me.

With the latest update there is now a NPM explorer, https://code.visualstudio.com/updates/v1_23#_npm-script-running.

I think it could be improved, but it is a really good start!

**Comment by dbaeumer** [comment]
I will rename the issue since the major item left in the issue discussed is the task runner viewlet / UI.

Note that with the new API this might be implementable by an extension (see the npm script runner viewlet)

**Comment by connorshea** [comment]
🙏 I'd love to see a sidebar like this in VS Code

**Comment by heartacker** [comment]
please give us  viewlet
